lighthouse-core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js

```diff
@@ -20,162 +19,184 @@
    */
-  static findBlockingTags(driver, networkRecords) {
-    const scriptSrc = `(${collectTagsThatBlockFirstPaint.toString()}())`;
+  static async findBlockingTags(driver, networkRecords) {
     const firstRequestEndTime = networkRecords.reduce(
       (min, record) => Math.min(min, record.endTime),
       Infinity
     );
-    return driver.evaluateAsync(scriptSrc).then(tags => {
-      const requests = TagsBlockingFirstPaint._filteredAndIndexedByUrl(networkRecords);
-
-      return tags.reduce((prev, tag) => {
-        const request = requests[tag.url];
-        if (request && !request.isLinkPreload) {
-          // Even if the request was initially blocking or appeared to be blocking once the
-          // page was loaded, the media attribute could have been changed during load, capping the
-          // amount of time it was render blocking. See https://github.com/GoogleChrome/lighthouse/issues/2832.
-          const timesResourceBecameNonBlocking = (tag.mediaChanges || [])
-            .filter(change => !change.matches)
-            .map(change => change.msSinceHTMLEnd);
-          const earliestNonBlockingTime = Math.min(...timesResourceBecameNonBlocking);
-          const lastTimeResourceWasBlocking = Math.max(
-            request.startTime,
-            firstRequestEndTime + earliestNonBlockingTime / 1000
-          );
-
-          prev.push({
-            tag,
-            transferSize: request.transferSize || 0,
-            startTime: request.startTime,
-            endTime: Math.min(request.endTime, lastTimeResourceWasBlocking),
-          });
-
-          // Prevent duplicates from showing up again
-          requests[tag.url] = null;
-        }
-
-        return prev;
-      }, []);
-    });
+    const tags = await driver.evaluate(collectTagsThatBlockFirstPaint, {args: []});
+    const requests = TagsBlockingFirstPaint._filteredAndIndexedByUrl(networkRecords);
+
+    /** @type {Array<LH.Artifacts.TagBlockingFirstPaint>} */
+    const result = [];
+    for (const tag of tags) {
+      const request = requests.get(tag.url);
+      if (!request || request.isLinkPreload) continue;
+
+      let endTime = request.endTime;
+      if (tag.tagName === 'LINK') {
+        // Even if the request was initially blocking or appeared to be blocking once the
+        // page was loaded, the media attribute could have been changed during load, capping the
+        // amount of time it was render blocking. See https://github.com/GoogleChrome/lighthouse/issues/2832.
+        const timesResourceBecameNonBlocking = tag.mediaChanges
+          .filter(change => !change.matches)
+          .map(change => change.msSinceHTMLEnd);
+        const earliestNonBlockingTime = Math.min(...timesResourceBecameNonBlocking);
+        const lastTimeResourceWasBlocking = Math.max(
+          request.startTime,
+          firstRequestEndTime + earliestNonBlockingTime / 1000
+        );
+        endTime = Math.min(endTime, lastTimeResourceWasBlocking);
```

Comment:
this section is kind of weird, because `tag.mediaChanges` is usually an empty array, so `timesResourceBecameNonBlocking` is too, and `Math.min(...[])` is `Infinity`. However whenever it's `Infinity` it always loses the _next_ `Math.min` with `request.endTime`, so it does work reliably, so I didn't change anything. Happy to go the other way, though.

Reply:
> this section is kind of weird, because `tag.mediaChanges` is usually an empty array, so `timesResourceBecameNonBlocking` is too, and `Math.min(...[])` is `Infinity`. However whenever it's `Infinity` it always loses the _next_ `Math.min` with `request.endTime`, so it does work reliably, so I didn't change anything. Happy to go the other way and make the `[]` case more explicitly handled, though.

I ended up making this an explicit check for an empty array so the next person doesn't have to figure out how it's not broken :)